Resolve data_files by split name (#3221)
huggingface/datasets: docs/source/repository_structure.rst

Structure your repository
=========================

To host and share your dataset, you can create a dataset repository on the Hugging Face Dataset Hub and upload your data files.

> This guide aims to explain what structure is expected when you upload a dataset.

Reviewer: ```suggestion
This guide will show you how to structure your dataset repository when you upload it.
``````

> A dataset with a supported structure can be loaded automatically using ``load_dataset``, and also has a nice preview on its dataset page on the Hub.

Reviewer: ```suggestion
A dataset with a supported structure can be loaded automatically with ``load_dataset``, and it will have a preview on its dataset page on the Hub.
```

.. Note::

>     In the following examples, CSV files are used. However you can use any supported format (text, JSON, JSON Lines, CSV, Parquet).

Reviewer: ```suggestion
    The following examples use CSV files, but you can use any supported format (text, JSON, JSON Lines, CSV, Parquet).
```

Main use-case
-------------

> Here is an easy structure that you can use for your repository. It contains two files: `train.csv` and `test.csv`.

Reviewer: ```suggestion
The simplest dataset structure two files: `train.csv` and `test.csv`.
```

> It also contains the dataset card `README.md` that is the `dataset card <dataset_card.html>`__ that is displayed on the dataset webpage.

Reviewer: ```suggestion
Your repository will also contain a `README.md` file, the `dataset card <dataset_card.html>`__ displayed on your dataset page.
```

::

    my_dataset_repository/
    ├── README.md
    ├── train.csv
    └── test.csv

Splits and file names
---------------------

> The train/validation/test splits of your dataset are automatically infered from the file names.

Reviewer: ```suggestion
🤗 Datasets automatically infers the train/validation/test splits of your dataset from the file names. All the files that contain **train** in their names are considered part of the train split. The same idea applies to the test and validation split: 
* All the files that contain **test** in their names are considered part of the test split.
* All the files that contain **valid** in their names are considered part of the validation split.
```

All the files that contain "train" in their names are considered part of the train split.
This is the same for the test and validation split: all the files that contain "test" (resp. "valid") in their names are considered part of the test (resp. validation) split.
Here is an example where all the files are placed into a directory named `data`:

::

    my_dataset_repository/
    ├── README.md
    └── data/
        ├── train.csv
        ├── test.csv
        └── valid.csv

Multiple files per split
------------------------

> If your train is split into several files it also works.

Reviewer: ```suggestion
If one of your splits comprises several files, 🤗 Datasets can still infer whether it is the train/validation/ test split from the file name. For example, if your **train** split spans several files:
```

::

    my_dataset_repository/
    ├── README.md
    ├── train_0.csv
    ├── train_1.csv
    ├── train_2.csv
    ├── train_3.csv
    ├── test_0.csv
    └── test_1.csv

> Just make sure that all the files of your train set have "train" inside their names (same for test and validation).

Reviewer: ```suggestion
Just make sure that all the files of your **train** set have **train** in their names (same for test and validation). It doesn’t matter if you add a prefix or suffix to **train** in the file name (like `my_train_file_00001.csv`, for example). 🤗 Datasets can still infer the appropriate split.
```

It doesn't matter if you add a prefix before "train" or a suffix after "train" in the file name (like "my_train_file_00001.csv" for example).

> For convenience you can also place your data files into different directories. In this case the split name is infered from the directory name.

Reviewer: ```suggestion
For convenience, you can also place your data files into different directories. In this case, the split name is inferred from the directory name.
```

::

    my_dataset_repository/
    ├── README.md
    └── data/
        ├── train/
        │   ├── shard_0.csv
        │   ├── shard_1.csv
        │   ├── shard_2.csv
        │   └── shard_3.csv
        └── test/
            ├── shard_0.csv
            └── shard_1.csv

Custom split names
------------------

> If you have other splits in addition to the traditionnal train/validation/test, you must use the following structure.

Reviewer: ```suggestion
If you have other data files in addition to the traditional train/validation/test sets, you must use the following structure. Follow the file name format exactly for this type of structure: `data/<split_name>-xxxxx-of-xxxxx.csv`. Here is an example with three splits: **train**, **test**, and **random**:
```

Note that in this case you must follow the file name format exactly: "data/<split_name>-xxxxx-of-xxxxx.csv".
Here is an example with three splits "train", "test" and "random":

::

    my_dataset_repository/
    ├── README.md
    └── data/
        ├── train-00000-of-00003.csv
        ├── train-00001-of-00003.csv
        ├── train-00002-of-00003.csv
        ├── test-00000-of-00001.csv
        ├── random-00000-of-00003.csv
        ├── random-00001-of-00003.csv
        └── random-00002-of-00003.csv

Multiple configuration (WIP)
----------------------------

> You can separate different configurations of your dataset (for example if it is splti in different languages) by using one directory per configuration.

Reviewer: ```suggestion
You can specify different configurations of your dataset (for example, if a dataset contains different languages) with one directory per configuration.
```
